src/cls/rbd/cls_rbd_types.h

```diff
@@ -674,20 +693,20 @@ struct SnapshotNamespace : public SnapshotNamespaceVariant {
   void decode(ceph::buffer::list::const_iterator& it);
```

Comment:
(this comment is for the line 669, `SnapshotNamespace(T&& t` constructor)

So, as I understand your "test/rbd_mirror: use ASSERT_TRUE() to compare iterators" commit log message, this constructor does not work any more.

Is there any way to actually make it work? Then it seems we would not need changing the tests to use `ASSERT_TRUE` (I don't mind of using `ASSERT_TRUE`, but I am afraid someone, may be me, will hit this again when adding new tests and will have to spend time figuring out why it does not work).

And if it is not possible to make the constructor work, then may be it is better to remove it? Or may be I completely misunderstood the issue?

Reply:
@trociny hi Mykola, thanks for your review.

the reason why i wanted to replace `boost::variant<>` with `std::variant<>` was, as the commit message explains, `boost::variant` does not allow us to compare a `boost::variant` with its derived class. i think the purpose of `SnapshotNamespace(T&& t)` is to convert a type in the "union" to `SnapshotNamespace`. for instance, this template constructor allows us to construct a `SnapshotNamespace` from `UserSnapshotNamespace`.

the `ASSERT_TRUE()` fix is all about the printability of iterators. it seems that the templated constructor looks like a honey pot from the compilers perspective, and it just insists on using it when looking up the candidates of `operator<<(..)`. that's why it fails to fallback to the fallback printer provided by GTest.

anyhow, i changed the way how the constructor is implemented. seems it's a better solution

- no need to have the `ASSERT_TRUE()` change
- the constructor is also simpler.